flow/BlobCipher.cpp

```diff
@@ -0,0 +1,392 @@
+}
+
+EncryptBlobCipherAes265Ctr::~EncryptBlobCipherAes265Ctr() {
+	EVP_CIPHER_CTX_free(ctx);
+}
+
+// DecryptBlobCipher class methods
+
+DecryptBlobCipherAes256Ctr::DecryptBlobCipherAes256Ctr(Reference<BlobCipherKey> key, const BlobCipherIV& iv)
+  : ctx(EVP_CIPHER_CTX_new()) {
+	if (ctx == nullptr) {
+		throw internal_error();
+	}
+	if (!EVP_DecryptInit_ex(ctx, EVP_aes_256_ctr(), nullptr, nullptr, nullptr)) {
+		EVP_CIPHER_CTX_free(ctx);
+		throw internal_error();
+	}
+	if (!EVP_DecryptInit_ex(ctx, nullptr, nullptr, key.getPtr()->data(), iv.data())) {
+		EVP_CIPHER_CTX_free(ctx);
+		throw internal_error();
+	}
+}
+
+StringRef DecryptBlobCipherAes256Ctr::decrypt(unsigned char const* ciphertext,
+                                              const int ciphertextLen,
+                                              const BlobCipherEncryptHeader& header,
+                                              Arena& arena) {
+	TEST(true); // Decrypting data with BlobCipher
+	// validate header flag sanity; for now only one valid header version exists.
+	if (header.flags.headerVersion != EncryptBlobCipherAes265Ctr::ENCRYPT_HEADER_VERSION ||
+	    header.flags.encryptMode != BLOB_CIPHER_ENCRYPT_MODE_AES_256_CTR) {
+		throw internal_error();
+	}
+	// encrypted byte checksum sanity; protection against data bit-rot/flip.
+	BlobCipherChecksum checksum = XXH3_64bits(ciphertext, ciphertextLen);
+	if (checksum != header.checksum) {
+		throw internal_error();
+	}
+	auto plaintext = new (arena) unsigned char[ciphertextLen + AES_BLOCK_SIZE];
+	int bytesDecrypted{ 0 };
+	if (!EVP_DecryptUpdate(ctx, plaintext, &bytesDecrypted, ciphertext, ciphertextLen)) {
+		EVP_CIPHER_CTX_free(ctx);
+		throw internal_error();
+	}
+	int finalBlobBytes{ 0 };
+	if (EVP_DecryptFinal_ex(ctx, plaintext + bytesDecrypted, &finalBlobBytes) <= 0) {
+		EVP_CIPHER_CTX_free(ctx);
+		throw internal_error();
+	}
+	return StringRef(plaintext, bytesDecrypted + finalBlobBytes);
+}
+
+DecryptBlobCipherAes256Ctr::~DecryptBlobCipherAes256Ctr() {
+	EVP_CIPHER_CTX_free(ctx);
+}
+
+// HmacSha256DigestGen class methods
+
+HmacSha256DigestGen::HmacSha256DigestGen(const unsigned char* key, size_t len) : ctx(HMAC_CTX_new()) {
+	if (!HMAC_Init_ex(ctx, key, len, EVP_sha256(), nullptr)) {
+		throw internal_error();
+	}
+}
+
+HmacSha256DigestGen::~HmacSha256DigestGen() {
+	HMAC_CTX_free(ctx);
+}
+
+StringRef HmacSha256DigestGen::digest(const unsigned char* data, size_t len, Arena& arena) {
+	TEST(true); // Digest generation
+	unsigned int digestLen = HMAC_size(ctx);
+	auto digest = new (arena) unsigned char[digestLen];
+	if (HMAC_Update(ctx, data, len) != 1) {
+		throw internal_error();
+	}
+	if (HMAC_Final(ctx, digest, &digestLen) != 1) {
+		throw internal_error();
+	}
+	return StringRef(digest, digestLen);
+}
+
+// Only used to link unit tests
+void forceLinkBlobCipherTests() {}
+
+// Test BlobCipherKey caching mechanism, tests cases includes:
+// 1. Insert & retrieval of latest cipher for multiple encryption domains
```

Comment:
Should we add a unit test for encryption/decryption correctness as well? That the input matches the decrypted data?

Reply:
EncryptionOps test validates the correctness of the encyrption/decryption operations as mentioned in the above comment.